File: 21558 - Nithin/week 7/82-remove-duplicates-from-sorted-list-ii/82-remove-duplicates-from-sorted-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(head == nullptr || head->next == nullptr)
            return head;
        unordered_map<int, int> map; 
        ListNode* temp = head;
        
        while(temp != NULL){
            map[temp->val]++;
            temp = temp->next;
        }
        
        ListNode* temp2 = head;
        ListNode* temp3 = head;
        while(temp2 != NULL){
            if(map[temp2->val] > 1){
                for(int i = 0;i < map[temp2->val] -1;i++){
                    temp2->next = temp2->next->next;
                }
                if(temp2 == temp3){
                    head = temp2->next;
                    temp3 = head;
                }
    
                else
                {
                    while(temp3->next != temp2)
                       temp3 = temp3->next;
                    temp3->next = temp2->next;
                }
            }
            temp2 = temp2->next;
        }
        return head;
    }
};

File: 21561 - Suvam Tamang/Week 7/82_Remove Duplicates from Sorted List II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if(head == null)
            return null;
        if(head.next == null)
            return head;
        ListNode dummyHead = new ListNode(0);
        dummyHead.next = head;
        ListNode temp = dummyHead;
        ListNode curr = head;
        while(curr != null && curr.next != null) {
            while(curr != null && curr.next != null && curr.val == curr.next.val)
               curr = curr.next;
            if(temp.next == curr)
                temp = curr;
            else
                temp.next = curr.next;
            curr = curr.next;
        }
        return dummyHead.next;
    }
}
